---
title: "STA3020 - Assignment 4"
author: "Chesia Anyika - 665567"
date: "2024-03-14"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Question

Canonical Correlation Analysis (CCA) is a statistical technique used to explore the relationships between two sets of variables. Given two multivariate datasets, X and Y, CCA aims to identify linear combinations of variables from each dataset that exhibit the highest correlation. However, when dealing with high-dimensional data where the number of variables in X and/or Y is very large,traditional CCA methods require modifications to handle the increased complexity.

The paper authored by Gonzalez et al. (2008) introduces a regularized CCA approach (outlined insection 2.4) specifically designed for high-dimensional datasets. While this method is suggested, alternatives from existing literature may also be explored.

This homework assignment focuses on implementing both classical CCA and its high-dimensional counterpart.

− The task involves applying the chosen approach to the nutrimouse dataset, accessible through the CCA R package. Detailed information regarding the dataset can be found in the corresponding paper.

− For visualization, a two-dimensional graph is generated by examining the first two dimensions resulting from CCA.

− The assignment is to be completed individually, with the submission of a report written in RMarkdown format. The report should encompass:

▪ A brief overview of both classical and high-dimensional CCA methodologies implemented.

▪ Application of the preferred high-dimensional CCA method to the nutrimouse dataset.

▪ Classical CCA performed on multivariate data with p \< n. (Note: Due to limitations, classical CCA cannot be applied directly to the entire X data matrix. Data should be subsetted to reflect the scenario of p \< n.)

▪ High-dimensional CCA conducted on data with p \> n.

▪ Interpretation and conclusions drawn from the data analysis results.

# Libraries

```{r}
library(CCA)
library(CCP)
```

# Execution

## Data Preparation and Exploratory Analysis

```{r}
#import the required data
data("nutrimouse")

#get a description of the data
?nutrimouse

#get a summary of the data
summary(nutrimouse)
```

The description of the data gives the following information:

The `nutrimouse` dataset comes from a nutrition study in the mouse. It was provided by Pascal Martin from the Toxicology and Pharmacology Laboratory (French National Institute for Agronomic Research).

**Format**

The dataset is a list that comprises of two sets of variables were measured on 40 mice:

-   `gene`: A data frame of dimensions 40 by 120, with numerical variables. This represents the expressions of 120 genes potentially involved in nutritional problems.

-   `lipid`: A data frame of dimensions 40 by 21, with numerical variables. This represents concentrations of 21 hepatic fatty acids.

The 40 mice were distributed in a 2-factors experimental design each with four 4 replicates for the following two factor variables:

-   `diet`: A factor vector of length 40, with two levels:

    -   `wt` - represents wild-type genotypes

    -   `ppar` - represents PPARalpha genotypes.

-   `genotype`: A factor vector of length 40 which represents the different oils used for experimental diets preparation, with 5 levels:

    -   `ref` - corn and colza oils (50/50) for a reference diet.

    -   `coc` - hydrogenated coconut oil for a saturated fatty acid diet

    -   `fish` - corn/colza/enriched fish oils in the ratio 43/43/14 for the fish diet.

    -   `lin` - linseed oil for an Omega3-rich diet

    -   `sun` - sunflower oil for an Omega6 fatty acid-rich diet

Thus each combination of `genotype` and `diet` is replicated four times. In other words, there are four mice for each unique combination of `genotype` and `diet`. Replication helps in improving the reliability and robustness of the experimental results by reducing the impact of random variation.

The levels of the factor variables can be confirmed as follows:

```{r}
#levels of the diet variable
levels(nutrimouse$diet)

#levels of the genotype variable
levels(nutrimouse$genotype)
```

## Canonical Correlation Analysis (CCA)

> **Brief Definition**
>
> Canonical Correlation Analysis (CCA) is a multivariate statistical technique used to explore the relationships between two sets of variables, typically in matrix form and referred to as $X$ and $Y$. It aims to identify linear combinations of variables from each set in such a way that the correlation between these combinations, called [canonical variates]{.underline}, is maximised.

### Classical CCA

> If we consider the two matrices $X$ and $Y$ are of order $n \times p$ and $n \times q$ respectively, Classical CCA assumes that $p \leq n$ and $q \leq n$. Thus, classical CCA is suitable for data-sets in which the number of observations ( $n$ ) do not exceed the number of variables ( $p, q$ ) for both matrices.

I decided to use the `gene` and `lipid` data-frames in order to assess correlations between different types of genes potentially involved in nutritional problems, and the concentrations of 21 hepatic fatty acids in the 40 observed mice.

First, I defined my matrices with rows as observations and columns as variables.

```{r}
#define matrices X and Y
X <- as.matrix(nutrimouse$gene)
Y <- as.matrix(nutrimouse$lipid)
```

As the two data-frames use different units of measurement, one being for genes and the other lipid concentration, Standardisation is required.

> Standardisation, or Z-Score Normalisation refers to data a preprocessing technique used to transform variables so that they have a mean of 0 and a standard deviation of 1. It is useful for better comparability and interpretability of results, as it allows for direct comparison of variables with different scales of measurment.
>
> **Formula**:
>
> $$z_i = \frac{x_i - \mu}{\sigma}$$ Where:
>
> -   $z_i$​ is the standardized value of observation $x_i$.
>
> -   $x_i$​ is the original value of observation $x_i$.
>
> -   $\mu$ is the mean of variable $X$.
>
> -   $\sigma$ is the standard deviation of variable $X$.

In r, this is computed using the `scale()` function. I concatenated my two matrices before scaling, to maintain the covariance structure between them. I then separated them and obtained the dimenstions of the resultant matrices, as follows:

```{r}
#concatenate the matrices column-wise
XY <- cbind(X,Y)

#standardize the combined matrix
XY.std <- scale(XY)

#split the standardized matrix back into two separate matrices
X.std <- XY.std[, 1:ncol(X)]
Y.std <- XY.std[, (ncol(X) + 1):ncol(XY)]

#get the dimensions of the matrices
dim(X.std)
dim(Y.std)
```

The dimensions obtained show that $X_{std}$ is a $40 \times 120$ matrix while $Y_{std}$ is a $40 \times 21$ matrix. This is not workable for Classical Canonical correlation as matrix $X_{std}$ violates the condition that $p \leq n$, the variables should not exceed the observations. Thus, we can subset the data to run CCA on a random sample of 10 variables (genes) instead of the full 120.

```{r}
#subset the X matrix
X.sub <- X.std[, sample(1:120, size = 10)]

#get the dimensions of the matrix
dim(X.sub)
```

Now, the order of matrix $X_{sub}$ is $40 \times 10$, which is suitable for CCA.

I ran CCA using the `cc()` function from the `CCA` package on $X_{sub}$ and $Y$, as follows:

```{r}
res.cc <- cc(X.sub, Y)
```

Examining some aspects of the output

```{r}
res.cc$cor
```

These are the canonical covariate coefficients computed with the first coefficient (0.9766973) representing the strength of association between the first pair of canonical variates, and so on. This suggests an overall high positive association between gene expression and concentration of lipids.

#### Visualisations

I visualised my results in a barplot as shown:

```{r}
barplot(res.cc$cor, xlab = "Dimension", ylab = "Canonical correlations", names.arg = 1:10, ylim = c(0,1))
```

The barplot shows a general trend of high positive association between gene expression and the concentration of lipids, with only the 9th and 10th coefficients being below 0.6.

I also visualised my resuts as shown:

```{r}
plt.cc(res.cc,var.label = TRUE, ind.names = paste(nutrimouse$genotype, nutrimouse$diet, sep = "-"))
```

### Regularised CCA (RCCA)

> If we consider the two matrices $X$ and $Y$ are of order $n \times p$ and $n \times q$ respectively, Classical CCA can not be performed if the number of observations ( $n$ ) is less than the greatest amount of variables in both data set, defined as $n ≤ max(p, q)$. In this case, a specialised method for Higher Dimensional datasets is required, and we use Regularised CCA which is suitable for data-sets in which the number of observations exceeds the number of variables for at least one matrix.
>
> **Regularisation**\
> In the context of RCCA, regularization modifies the covariance matrices $S_{XX}$​ and $S_{YY}$​ by adding regularization terms controlled by parameters $\lambda_1$​ and $\lambda_2$ respectively. These modified covariance matrices are defined as follows:
>
> $$ \Sigma_{XX}(\lambda1) = S_{XX} + \lambda_1I_p\\
> \Sigma_{YY}(\lambda_2) = S_{YY} + \lambda_2I_p$$
>
> where:
>
> -   $S_{XX}$ and $S_{YY}$ are the sample covariance matrices of the variables $X1,…,Xp$ and $Y1​,…,Yq​$ respectively.
>
> -   $I_p$​ and $I_q$​ are identity matrices of dimensions $p \times p$ and $q \times q$ respectively.
>
> -   $\lambda_1$ and $\lambda_2$ are regularization parameters that control the degree of regularization applied to the covariance matrices.

For RCCA, I first determined appropriate values for the regularisation parameters, $\lambda_1$ and $\lambda_2$ using the `estim.regul()` function on the standardised matrices.

```{r}
res.regul <- estim.regul(X.std, Y.std)
```

My obtained regularisation parameters are $\lambda_1 = 0.5005$ and $\lambda_2 = 0.001$. The regularisation parameters should maximise on the CV Score, which stands for the cross validation score, and runs on a scale from $0 - 1$

The CV Score obtained is $0.9698$ to 4 significant figures. This is an excellent score, thus the regularisation parameters are suitable for use in the RCCA.

I then stored the obtained regularisation parameters in the variables `lambda1` and `lambda2` , as well as the CV score in the variable `CV_score` .

```{r}
lambda1 <- res.regul$lambda1
lambda2 <- res.regul$lambda2
CV_score <- res.regul$CVscore

# Print out the results
cat("Optimal values:\n")
cat("lambda1 =", lambda1, "\n")
cat("lambda2 =", lambda2, "\n")
cat("CV-score =", CV_score, "\n")
```

I then performed RCCA using the `rcc` fundtion from the CCA library, with the chosen regularisation parameters

```{r}
res.rcc <- rcc(X, Y, lambda1, lambda2)
```

```{r}
res.rcc$cor
```

This output has more correlation coefficients than in CCA, due to the larger matrix used. The first regularised canonical covariate coefficient is 0.65, representing the association between the first pair of canonical covariates . This suggests that there is a moderate association between gene expression and the concentration of lipids.

#### Visualisations

I visualised my results in the barplot as follows:

```{r}
barplot(res.rcc$cor, xlab = "Dimension", ylab = "Canonical correlations", names.arg = 1:21, ylim = c(0,1))
```

The barplot shows a general trend of moderate positive association between gene expression and concentration of lipids, with only the first, second and third coefficients having coeffiecients greater than 0.4.

I also visualised my results as follows:

```{r}
plt.cc(res.rcc, var.label = TRUE, ind.names = paste(nutrimouse$genotype, nutrimouse$diet, sep = "-"))
```

# Conclusion

As per the CCA and the RCCA, it can be concluded that gene expression and concentration of lipids have an overall positive correlation, with the positive association being greater using subset `genes` data for the CCA, and the association being moderate when using the whole `genes`.
